
August 2015¶
Modified September 2016
Author: FBB
reproduce the result in
http://www.mdrc.org/sites/default/files/What%20Strategies%20Work%20for%20the%20Hard%20FR.pdf

In [22]:
import os
import sys
import numpy as np
import pylab as pl
import scipy.stats as sc


import json
import os
#s = json.load( open(os.getenv("PUI2015")+"/fbb_matplotlibrc.json") )
#pl.rcParams.update(s)

%pylab inline

#from IPython.display import Image
#Image(filename='../plotsforclasses/NYCReentryprogram_title.png')
print('Is this working')

Populating the interactive namespace from numpy and matplotlib
Is this working


NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05
$H_0: P_0 - P_1 \geq$ 0
$H_a: P_0 - P_1 $< 0
$\alpha$ = 0.05
this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above):¶
$P_0=0.035, P_1=0.701$

In [5]:
alpha=0.05
# we like fractions better then percentages. as a rule of thumb, either use fractions or counts
P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0-P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

n_0 = 409
n_1 = 564

#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

we must assess the statistical significance



2 samples, categorical data¶
TWO OPTIONS z test, or chi-square test.
START WITH Z TEST
the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the _assumption of normality
must define the sample standard deviation

In [6]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.421047276465


standard deviation of the sampling distribution the distribution is Binomial, the binomail stdev is
(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!):
$\sqrt{\frac{p(1 - p)}{n}}$
for 2 samples this becomes
$\sqrt{ \frac{ \hat{p}(1 - \hat{p})} {n0} + \frac{ \hat{p}(1 - \hat{p})} {n1} }$
cfr: page 138 of Statistics in a Nutshell, eq. 5.12 and here http://stattrek.com/hypothesis-test/difference-in-proportions.aspx?Tutorial=AP
$\hat{p} =\frac{p_0  n_0 + p_1  n_1}{n_0+n_1}$

In [7]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1)) 

In [8]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)

20.7697865408



note that using p0-p1 or p1-p0 at the numerator is equivalent because the standardizes normal value of z has mean 0 (see image below) so that we can use the absolute value of the z score, or equivalently look for $P[Z&lt;z]$ if z is positive, and $P[Z&gt;z]$ if z is negative.

In [9]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9998


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


!!!!!!!!TODO FOR YOU
what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?¶
Null hypothesis?
$H_0$?
$H_a$?
$P_0 = ??, P_1= ??$



Null Hypothesis: THe ercentage of former prisoners convicted of a felony in years 1-3 after relase is the same or higher for inmates who went through the program as for those who did not go through the program, at a signifigance level of .05.  


Ha: The percentage of prisoners who went through the program who were convicted of a felony in years 1-3 is lower than the prisoners who did not go through the program 


$H_0 = P_0 - P_1 =< 0    alpha = .05$
$H_a = P_0 - P_1 > 0$

$P_0 = .117 , P_1 = .10$

n_0 = 409
n_1 = 564

#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1


look up data table and insert the appropriate values to get the appropriate result! you can use the functions I defined above, with different arguments

In [11]:
P_0 = .117
P_1 = .10
n_0 = 409
n_1 = 564

if P_0-P_1 <= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

we must assess the statistical significance


In [12]:
z_3y = zscore(P_0, P_1, se(p(P_0, P_1, n_0, n_1), n_0, n_1 ))
z_3y

0.84628298260493418

In [13]:
#z = .7995
alpha=.05

p_3y = 1 - .7995


def report_result(p,a):
    print ('is the p value {0:.2f} smaller than the critical value {1:.2f}? '.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format( 'rejected' if p < a  else 'not rejected') )

    
report_result(p_3y,alpha)

is the p value 0.20 smaller than the critical value 0.05? 
NO!
the Null hypothesis is not rejected


Now lets do it with the $\chi^2$ test
this analysis can also be done with the $\chi^2$ test, and the  $\chi^2$ distribution,
see A. Marengo on how to choose a test and "Statistics In a Nutshell Chapter 4", or http://math.hws.edu/javamath/ryan/ChiSquare.html (if you are really just interested in the formula at face value)
The chisq statistics tests the statistics calculated as :
$\chi^2 = \sum_{i} \frac{(observation_i - expectation_i)^2}{expectation_i}$
against a chi sq distribution.
If we talk about sample fractions that is
$\chi^2 = \sum_i \frac{(f_{i,observed} - f_{i,expectated})^2}{f_{i,expected}}$
Where i indicates the sum over each cell.
turns out this quantity is distributed according to a chi square distribution, so if i get the $\chi^2$ statistics i can compare it to the full chisq distribution and see how far in the tail it is
The trickiest part (but not that tricky) is to figure out how to construct the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example: Thisis called a CONTINGENCY TABLE

The trickiest part (but not that tricky) is to figure out how to construct the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example: Thisis called a CONTINGENCY TABLE
success	failure	
test sample	number of successes in test	number of failures in test	number members of test sample
control sample	number of successes in control	number of failures in control	number members of control sample
total successes	total failures	number of all members
employed in subsadized job	employed	not employed
test sample	$0.701*564$	$0.299*564$	564
control sample	$0.035*409$	$0.965*4.09$	409
total	409.679	562.912	973
for each cell the expected ratio is the product of the total of all rows and all columns, devided by the total
FOR EACH CELL:
expected_cell = $\frac{row~tot * col~tot}{total}$


!!!!!! TODO FOR YOU!
lets see what the chi sq statistics says about the conviction for fellonies (row 10)¶
fill in the contingency table below
$convicted of a fellony	yes	no
test sample			
control sample			
total

| COnvictd of a felony  | yes  | no  |   |   |
|---|---|---|---|---|
|test sample   | .1*568  | .9*568  | 568  |   |
| control sample  | .117*409  |.883*409   | 409  |   |
|  total | 104.653  | 872.347  |  977 |   |

In [29]:
sample_values = np.array([[0.1 * 568, 0.9 * 568], [0.117 * 409, .883 * 409]])
#print (evalChisq(sample_values)

#THis is what should go into the equation
       
 

In [31]:
import numpy as np

def evalChisq(values):
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()


In [32]:
evalChisq(sample_values)

0.71849391750528868

#Chi square = approx .72 (rounded)
#With DOF =1.  The chisquare value appears to fall b/w .5 and .25 on the chart at
#http://passel.unl.edu/Image/Namuth-CovertDeana956176274/chi-sqaure%20distribution%20table.PNG

We thus cannot rejet the null hypotheis with an alpha of .05

I found this website  helpful
http://www2.lv.psu.edu/jxm57/irp/chisquar.html

